### PR TITLE
introduce the new integer types (u16, u32 & u256)

### PR DESCRIPTION
**Pull request by vgao1996**
## Overview
Support for the missing integer types (u16, u32 & u256) has landed in upstream Move. This PR brings them into `aptos-core` and makes other Move dependencies up to date.

## Backward Compatibility
This introduces a new gas schedule version (5) and a new feature flag `VM_BINARY_FORMAT_V6`. The VM will only start to accept the new binary format version if both conditions are met. 

## Work Items
- [x] Move fixes
- [x] api
- [x] sdk
  - [x] ts 
  - [x] python 
  - [x] rust (broken, won't fix) 
  - [x] go (broken, won't fix)
- [x] versioning/feature gating 
- [x] update bcs (won't land until this is addressed properly)
- [ ] check if the cli works
- [ ] check if the indexer works
- [ ] more tests

This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/5626)

### CODE REVIEW

File: api/types/src/derives.rs
Comment (gregnazario) [inline api/types/src/derives.rs:317]: Note, this needs to be cleaned up for u64 -> u256

File: network/src/protocols/wire/handshake/v1/mod.rs
Comment (gregnazario) [inline network/src/protocols/wire/handshake/v1/mod.rs:181]: Removing the with limit?

Comment (vgao1996) [reply network/src/protocols/wire/handshake/v1/mod.rs:181]: This is a temporary hack to make the code compile. There's a parallel task to fork the newer version of BCS and add back the with_limit variants. This PR will not land before we get that sorted out.

File: Cargo.toml
Comment (vgao1996) [inline Cargo.toml:293]: Note: need to point this to either an updated fork or upstream if we can get it patched timely.

**Comment by vgao1996** [comment]
@gregnazario IIRC once this PR lands, the cli will start to produce Move binaries with version 6, which is not compatible with the validator software we have on the current devnet, testnet and mainnet. Would this be a problem? The Move binary format serializer does have an option to output version 5 but it needs to be enabled.

**Comment by vgao1996** [comment]
@kent-white @CapCap could you help me verify if petra would work with this?

Same for @bowenyang007 with the indexer

Much appreciated!

File: aptos-move/framework/aptos-framework/doc/aptos_coin.md
Comment (clay-aptos) [inline aptos-move/framework/aptos-framework/doc/aptos_coin.md:7]: @vgao1996 , this can't be right, right? :-)

**Review by clay-aptos — DISMISSED** [review]
Hi Victor, I think the change to personal repo is unintentional.

File: aptos-move/framework/aptos-framework/doc/aptos_coin.md
Comment (vgao1996) [reply aptos-move/framework/aptos-framework/doc/aptos_coin.md:7]: This is indeed unintentional -- probably the artifacts of a rebase or search & replace). Thanks for catching it!

File: api/goldens/aptos_api__tests__transactions_test__test_get_transactions_returns_last_page_when_start_version_is_not_specified.json
Comment (movekevin) [inline api/goldens/aptos_api__tests__transactions_test__test_get_transactions_returns_last_page_when_start_version_is_not_specified.json:2068]: Why did this decrease?

File: api/types/src/derives.rs
Comment (movekevin) [inline api/types/src/derives.rs:268]: copy paste typo. Should be "JavaScript that do not parse **u32s** in JSON natively"

Comment (banool) [inline api/types/src/derives.rs:268]: How come we have these? Currently we only have U64 and up because JS can only handle U32 or lower natively. Meaning in this case, we should be able to just use the numbers directly instead of having this string wrapper type.

Comment (banool) [reply api/types/src/derives.rs:268]: Or perhaps we still need U32, but not U16.

File: aptos-move/framework/aptos-stdlib/sources/cryptography/bls12381.move
Comment (alinush) [inline aptos-move/framework/aptos-stdlib/sources/cryptography/bls12381.move:519]: Is this `location` field now mandatory with `expected_failure`? I couldn't find it in the Move book chapter on unit testing: https://move-language.github.io/move/unit-testing.html

File: api/goldens/aptos_api__tests__transactions_test__test_get_transactions_returns_last_page_when_start_version_is_not_specified.json
Comment (vgao1996) [reply api/goldens/aptos_api__tests__transactions_test__test_get_transactions_returns_last_page_when_start_version_is_not_specified.json:2068]: New compiler optimizations: https://github.com/move-language/move/commit/f04e00267dfb03c9efb11fb1b768c1cbc1341b8b

File: aptos-move/framework/aptos-stdlib/sources/cryptography/bls12381.move
Comment (vgao1996) [reply aptos-move/framework/aptos-stdlib/sources/cryptography/bls12381.move:519]: Good question. This is a new warning that was added in https://github.com/move-language/move/commit/926c4a1df234a373c4f11929f69be76c9581ea74. The compiler will raise warnings if there isn't a location field, which is fine by itself. However the unit testing framework currently treats warnings as test failures, which is not the desired behavior and we need to fix that separately.

File: aptos-move/framework/aptos-framework/doc/aptos_coin.md
Comment (vgao1996) [reply aptos-move/framework/aptos-framework/doc/aptos_coin.md]: Fixed

File: api/types/src/derives.rs
Comment (vgao1996) [reply api/types/src/derives.rs]: Thanks for catching us. I've deleted u16 and u32. JS uses float64 to represent numbers and should be able to handle integers between -2^53 and 2^53, so u32 should be fine.

File: ecosystem/typescript/sdk/src/aptos_types/transaction.ts
Comment (CapCap) [inline ecosystem/typescript/sdk/src/aptos_types/transaction.ts:425]: There’s a transaction argument vec<u8> above- do we not allow vectors of the other native types?

File: ecosystem/typescript/sdk/src/aptos_types/transaction.ts
Comment (CapCap) [inline ecosystem/typescript/sdk/src/aptos_types/transaction.ts:523]: Why do we do u128 first? Does this not truncate?

File: ecosystem/typescript/sdk/src/aptos_types/transaction.ts
Comment (vgao1996) [reply ecosystem/typescript/sdk/src/aptos_types/transaction.ts:459]: Here we're serializing the enum tag (6) in uleb128 format. The U32 is just an temporary in-memory representation of this tag.

File: ecosystem/typescript/sdk/src/aptos_types/transaction.ts
Comment (vgao1996) [reply ecosystem/typescript/sdk/src/aptos_types/transaction.ts:523]: Same as above, but I probably messed up the tags... Will get that fixed.

**Review by wrwg — APPROVED** [review]
I think we should better land this fast and then iterate as needed, also because the module renaming effort is underway.

File: api/goldens/aptos_api__tests__transactions_test__test_get_transactions_returns_last_page_when_start_version_is_not_specified.json
Comment (movekevin) [reply api/goldens/aptos_api__tests__transactions_test__test_get_transactions_returns_last_page_when_start_version_is_not_specified.json:2068]: Lol saving of 1 gas unit 😄

**Review by movekevin — APPROVED** [review]

File: ecosystem/typescript/sdk/src/aptos_types/transaction.ts
Comment (vgao1996) [reply ecosystem/typescript/sdk/src/aptos_types/transaction.ts:425]: Double checked this. We do not allow vectors of the other native types.

File: aptos-move/framework/aptos-framework/doc/aptos_coin.md
Comment (clay-aptos) [inline aptos-move/framework/aptos-framework/doc/aptos_coin.md:7]: I think this introduces a typo and resulting 404 ("langauge" should be "language"):
Correct - https://github.com/move-language/move/tree/main/language/documentation/tutorial
Incorrect - https://github.com/move-langauge/move/tree/main/language/documentation/tutorial

Comment (vgao1996) [reply aptos-move/framework/aptos-framework/doc/aptos_coin.md]: Fixed. Thank you so much for spotting this!!! I must be having a bad day...

File: crates/aptos/src/common/types.rs
Comment (gregnazario) [inline crates/aptos/src/common/types.rs:925]: This is related?

File: api/types/src/derives.rs
Comment (gregnazario) [inline api/types/src/derives.rs:163]: `u8`, `u16`, `u32` are integers right?

File: api/types/src/derives.rs
Comment (gregnazario) [inline api/types/src/derives.rs:165]: Similarly, this is u64, u128, u256

File: aptos-move/framework/aptos-framework/doc/aptos_coin.md
Comment (clay-aptos) [reply aptos-move/framework/aptos-framework/doc/aptos_coin.md]: My pleasure, my friend. I am happy to help.

File: crates/aptos/src/common/types.rs
Comment (vgao1996) [reply crates/aptos/src/common/types.rs:925]: Yes. By default, this flag is not set and the cli will tell the move compiler to use bytecode version 5 so to maintain compatibility with our current testnet and mainnet. The user can however, opt into version 6 using this flag if they want to try out the new integer types.

File: ecosystem/typescript/sdk/src/transaction_builder/builder_utils.test.ts
Comment (vgao1996) [reply ecosystem/typescript/sdk/src/transaction_builder/builder_utils.test.ts:161]: This is intentional. It's a test trying to parse an invalid type tag.

File: aptos-move/aptos-sdk-builder/src/golang.rs
Comment (vgao1996) [reply aptos-move/aptos-sdk-builder/src/golang.rs:727]: https://github.com/aptos-labs/aptos-core/issues/5837

**Pull request merged**
